src/gpu/locks.rs

```diff
@@ -14,31 +14,107 @@ use crate::gpu::{CpuGpuMultiexpKernel, GpuName};
 
 const GPU_LOCK_NAME: &str = "bellman.gpu.lock";
 const PRIORITY_LOCK_NAME: &str = "bellman.priority.lock";
-fn tmp_path(filename: &str) -> PathBuf {
+
+fn tmp_path(filename: &str, id: Option<UniqueId>) -> PathBuf {
     let mut p = std::env::temp_dir();
-    p.push(filename);
+    let mut tmpfile = filename.to_owned();
+    if let Some(id_str) = id {
+        tmpfile.push_str(&(".".to_owned() + &id_str.to_string()));
+    }
+    p.push(tmpfile);
     p
 }
 
+#[derive(Debug)]
+struct LockInfo<'a> {
+    file: Option<File>,
+    path: Option<PathBuf>,
+    devices: Vec<&'a Device>,
+}
+
 /// `GPULock` prevents two kernel objects to be instantiated simultaneously.
 #[allow(clippy::upper_case_acronyms)]
 #[derive(Debug)]
-pub struct GPULock(File);
-impl GPULock {
-    pub fn lock() -> GPULock {
-        let gpu_lock_file = tmp_path(GPU_LOCK_NAME);
-        debug!("Acquiring GPU lock at {:?} ...", &gpu_lock_file);
-        let f = File::create(&gpu_lock_file)
-            .unwrap_or_else(|_| panic!("Cannot create GPU lock file at {:?}", &gpu_lock_file));
-        f.lock_exclusive().unwrap();
+pub struct GPULock<'a>(Vec<LockInfo<'a>>);
+
+impl GPULock<'_> {
+    pub fn lock() -> Self {
+        if let Ok(val) = std::env::var("BELLPERSON_GPUS_PER_LOCK") {
+            match val.parse::<usize>() {
+                Ok(val) if val > 0 => {
+                    let devices = Device::all();
+                    info!(
+                        "BELLPERSON_GPUS_PER_LOCK == {}, try lock {}/{} gpus",
+                        val,
+                        val,
+                        devices.len(),
+                    );
+
+                    let mut locks = Vec::new();
+                    for (index, device) in devices.iter().enumerate() {
+                        let uid = device.unique_id();
+                        let path = tmp_path(GPU_LOCK_NAME, Some(uid));
+                        debug!("Acquiring GPU lock {}/{} at {:?} ...", index, val, &path);
+                        let file = File::create(&path).unwrap_or_else(|_| {
+                            panic!("Cannot create GPU {:?} lock file at {:?}", uid, &path)
+                        });
+                        if file.try_lock_exclusive().is_err() {
+                            continue;
+                        }
+                        debug!("GPU lock acquired at {:?}", path);
+                        locks.push(LockInfo {
+                            file: Some(file),
+                            path: Some(path),
+                            devices: vec![device],
+                        });
+                        if locks.len() >= val {
+                            break;
+                        }
+                    }
+
+                    return GPULock(locks);
+                }
+                Ok(val) if val == 0 => {
+                    info!("BELLPERSON_GPUS_PER_LOCK == 0, free to use gpus");
```

Comment:
Another idea for a different log message: what about "BELLPERSON_GPUS_PER_LOCK == 0, no locks acquired"